[WTF] WorkerPool::Worker should use a CheckedPtr instead of a CheckedRef for the WorkerPool reference

carlosgcampos opened the pull request:
#### 5491e26242436c0d3daddbb03717afea739057fb
<pre>
[WTF] WorkerPool::Worker should use a CheckedPtr instead of a CheckedRef for the WorkerPool reference
<a href="https://bugs.webkit.org/show_bug.cgi?id=302929">https://bugs.webkit.org/show_bug.cgi?id=302929</a>

Reviewed by Nikolas Zimmermann.

Since AtomicThread keeps a reference of itself while the thread is
running, it can happen that the WorkerPool is destroyed before the
thread actually finishes even if join was called and threadIsStopping is
already called. This can happen easily when the system is under heavy
load, as we have seen in the bots when running the layout tests.

Canonical link: <a href="https://commits.webkit.org/303468@main">https://commits.webkit.org/303468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35cfebfb91153c9640187dfa9b0ac42f6ca17eb7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132502 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4997 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43556 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140021 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84484 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5185 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4756 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101300 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68570 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135448 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82094 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1283 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83255 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124572 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112438 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142673 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131010 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4667 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37399 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109674 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4749 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4030 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109856 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27840 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3544 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114967 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58040 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4721 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33317 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163977 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4555 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68172 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42593 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4812 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4678 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WTF/wtf/WorkerPool.cpp:
```diff
@@ -64,26 +71,30 @@ class WorkerPool::Worker final : public AutomaticThread {
     void threadDidStart() final
     {
         Locker locker { *m_pool->m_lock };
+        ASSERT(m_pool);
```
You are already accessing the pool line before?

csaavedra commented on Source/WTF/wtf/WorkerPool.cpp:
```diff
@@ -64,26 +71,30 @@ class WorkerPool::Worker final : public AutomaticThread {
     bool shouldSleep(const AbstractLocker& locker) final
     {
+        ASSERT(m_pool);
         return m_pool->shouldSleep(locker);
```
Now that `m_pool` is not `const`, you need to protect it with a stack variable because `shouldSleep()` is not trivial. Something like

```
return CheckedPtr { m_pool }->shouldSleep(locker);
```

for example.

carlosgcampos committed:
[WTF] WorkerPool::Worker should use a CheckedPtr instead of a CheckedRef for the WorkerPool reference
https://bugs.webkit.org/show_bug.cgi?id=302929

Reviewed by Nikolas Zimmermann.

Since AtomicThread keeps a reference of itself while the thread is
running, it can happen that the WorkerPool is destroyed before the
thread actually finishes even if join was called and threadIsStopping is
already called. This can happen easily when the system is under heavy
load, as we have seen in the bots when running the layout tests.

Canonical link: https://commits.webkit.org/303468@main
